src/librustc_middle/query/mod.rs

```diff
@@ -1257,5 +1257,9 @@ rustc_queries! {
             eval_always
             desc { "looking up enabled feature gates" }
         }
+
+        query resolve_instance(key: (ty::ParamEnv<'tcx>, DefId, SubstsRef<'tcx>)) -> Option<ty::Instance<'tcx>> {
```

Comment:
Just spotted this: whenever you have a `ty::ParamEnv` and some thing referring the to same parameters it adds bounds in scope of, like `(DefId, SubstsRef)` here, the right thing to do is to use `ty::ParamEnvAnd<(DefId, SubstsRef)>`.

Doing so should result in better caching (as the bounds in scope aren't kept if they're not needed by the `SubtsRef`), and is more semantically meaningful.

Reply:
Will do this in a drive-by commit.